src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1427,126 +1433,407 @@ let CheckMultipleInputsSequential (ctok, checkForErrors, tcConfig, tcImports, tc
+
+    let sourceFiles: FileInProject array =
+        inputs
+        |> List.toArray
+        |> Array.mapi (fun idx (input: ParsedInput) ->
+            {
+                Idx = idx
+                FileName = input.FileName
+                ParsedInput = input
+            })
+
+    let filePairs = FilePairMap(sourceFiles)
+
+    let graph =
+        DependencyResolution.mkGraph tcConfig.compilingFSharpCore filePairs sourceFiles
+
+    let nodeGraph =
+        let mkArtificialImplFile n = NodeToTypeCheck.ArtificialImplFile n
+        let mkPhysicalFile n = NodeToTypeCheck.PhysicalFile n
+
+        /// Map any signature dependencies to the ArtificialImplFile counterparts,
+        /// unless the signature dependency is the backing file of the current (implementation) file.
+        let mapDependencies idx deps =
+            Array.map
+                (fun dep ->
+                    if filePairs.IsSignature dep then
+                        let implIdx = filePairs.GetImplementationIndex dep
+
+                        if implIdx = idx then
+                            // This is the matching signature for the implementation.
+                            // Retain the direct dependency onto the signature file.
+                            mkPhysicalFile dep
+                        else
+                            mkArtificialImplFile dep
+                    else
+                        mkPhysicalFile dep)
+                deps
+
+        // Transform the graph to include ArtificialImplFile nodes when necessary.
+        graph
+        |> Seq.collect (fun (KeyValue (fileIdx, deps)) ->
+            if filePairs.IsSignature fileIdx then
+                // Add an additional ArtificialImplFile node for the signature file.
+                [|
+                    // Mark the current file as physical and map the dependencies.
+                    mkPhysicalFile fileIdx, mapDependencies fileIdx deps
+                    // Introduce a new node that depends on the signature.
+                    mkArtificialImplFile fileIdx, [| mkPhysicalFile fileIdx |]
+                |]
+            else
+                [| mkPhysicalFile fileIdx, mapDependencies fileIdx deps |])
+        |> Graph.make
+
+    // Persist the graph to a Mermaid diagram if specified.
+    if tcConfig.typeCheckingConfig.DumpGraph then
+        tcConfig.outputFile
+        |> Option.iter (fun outputFile ->
+            let outputFile = FileSystem.GetFullPathShim(outputFile)
+            let graphFile = FileSystem.ChangeExtensionShim(outputFile, ".graph.md")
+
+            graph
+            |> Graph.map (fun idx ->
+                let friendlyFileName =
+                    sourceFiles[idx]
+                        .FileName.Replace(tcConfig.implicitIncludeDir, "")
+                        .TrimStart([| '\\'; '/' |])
+
+                (idx, friendlyFileName))
+            |> Graph.serialiseToMermaid graphFile)
+
+    let _ = ctok // TODO Use it
```

Comment:
How can/should this be used past calling `RequireCompilationThread` on it?

Reply:
We're breaking the assumption that checking happens on the compilation thread as there are many threads now.

Which should hopefully mean that it's ok to discard the token.

Reply:
(So the TODO is unnecessary. I'm happy to change in a separate PR)